### PR TITLE
feat(store-devtools): provide the ability to connect extension outside of Angular zone

### LINKED ISSUE #3839: Remove dev tools event listener from zone.js detection
### Which @ngrx/* package(s) are the source of the bug?

store-devtools

### Minimal reproduction of the bug/regression with instructions

- Add `StoreDevtoolsModule.instrument({})` to app.module.ts, just like in [the example](https://stackblitz.com/edit/ngrx-seed?file=src%2Fapp%2Fapp.module.ts)
- Install [Angular DevTools](https://chrome.google.com/webstore/detail/angular-devtools/ienfalfjdbdpebioblfackkekamfmbnh)
- Through Angular DevTools, start the profiler in the example
- The profiler will show a lot of change detection events caused by `Window.addEventListener:message`
- Remove `StoreDevtoolsModule.instrument({})`, repeat step 3, the change detection events won't show up anymore

### Minimal reproduction of the bug/regression with instructions

It's expected that the `StoreDevtools` doesn't trigger any change detection.
Change detection can be costly if any component do some kind of 'heavy work' like rendering multiple components with `ngForOf`, even if `StoreDevtools` logs only in development, it's still a big overhead for a development environment, depending on the size and complexity of the project.

### Versions of NgRx, Angular, Node, affected browser(s) and operating system(s)

NgRx: 15.4.0
Angular: 15.2.3
Node: 16.16
Browser: Microsoft Edge 112.0.1722.34 (64-bit)
Operating system: Windows 10

### I would be willing to submit a PR to fix this issue

- [X] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by arturovt**
This commit updates the extension connection setup and wraps it with `runOutsideAngular` to prevent running change detection too frequently. This ensures that change detection is only triggered when all asynchronous actions have been processed.

Closes #3839

### CODE REVIEW

**Review by timdeschryver — CHANGES_REQUESTED** [review]
It seems like with this change importing an existing state throws an error.
This should be resolved first, or this should be added to the docs.

```txt
reducer.ts:75  ERROR Error: Action '[Auth/API] Login Redirect' running outside NgZone. https://ngrx.io/guide/store/configuration/runtime-checks#strictactionwithinngzone
    at inNgZoneAssert_reducer.ts:11:13
    at serialization_reducer.ts:24:23
    at immutability_reducer.ts:11:23
    at utils.ts:139:14
    at computeNextEntry (reducer.ts:72:17)
    at recomputeStates (reducer.ts:122:9)
    at reducer.ts:533:24
    at StoreDevtools.liftedStateSubscription.liftedAction$.pipe.state (devtools.ts:96:38)
    at scanInternals.js:14:21
    at OperatorSubscriber._this._next (OperatorSubscriber.js:33:21)
handleError @ core.mjs:10072
computeNextEntry @ reducer.ts:75
recomputeStates @ reducer.ts:122
(anonymous) @ reducer.ts:533
StoreDev
[state.txt](https://github.com/ngrx/platform/files/12147295/state.txt)
tools.liftedStateSubscription.liftedAction$.pipe.state @ devtools.ts:96
```

You can reproduce this by running the example app and importing my export.

```cmd
npx nx serve example-app
```

**Comment by arturovt** [comment]
@timdeschryver thanks for checking, I haven't imported the state and haven't checked that case, I'll have a look. It actually should keep functioning seamlessly w/o breaking changes, so I'd fix it.

**Comment by arturovt** [comment]
@timdeschryver I'll post my thoughts aloud so we can discuss the concerns which I think about:

- this change may cause breaking issues in unit tests that implicitly depend on a specific number and sequence of change detections to pass their assertions. However, I am not sure if someone is unit testing the Redux plugin, which is available only in the browser. This means that we might eliminate the reason for this to be a breaking change.
- the `DevtoolsExtension` (which has the `createChangesObservable` method) is not exposed publicly to end users through the NgRx API, which means I'm not sure if someone would explicitly subscribe to `createChangesObservable` and be affected by the fact that subscriber events would be coming from outside the Angular zone.
- the `StoreDevtool` are exposed publicly. Its public properties, which may be read by end users, are `liftedState` and `state`. The `state` basically pipes the `liftedState`. The `liftedState` is obtained from `liftedStateSubject.asObservable()`, where `liftedStateSubject` always calls its `next` within the Angular zone, since we call `enterNgZone()` previously.

Wdyt?

**Comment by timdeschryver** [comment]
@arturovt thanks for the elaborate answer.

- I would assume no projects are using devtools to unit test their components (I don't think this will break e2e tests)
- I also assume no projects are interacting with the devtools within their "user code".

The question was more about if this change could break existing application because the result of this change is that it won't trigger the CD as much as before.
But, that's probably not the case, since most of the apps turn devtools off in production.

The other thing I was thinking about are applications that don't rely on NgZone.

**Comment by arturovt** [comment]
Apps which have the NgZone noopped won’t be affected since NgZone class is substituted with NoopNgZone which does nothing (just invokes callbacks directly).

**Comment by brandonroberts** [comment]
I do think that there is a possibility of this being an unexpected change in behavior from those who are including the Devtools in production, which I know there are cases of this out there.

That being said, we always recommend not doing this, so we could be covered in that case.

**Comment by brandonroberts** [comment]
> 

I don't think that's necessary, but we could add that to the docs to be more clear. The other option would be to have this behind a flag, such as:

```ts
provideStoreDevtools({
  ngZone: 'enabled' | 'noop'
});
```

And have it default to `enabled` for now, and you can change it to `noop`. In the future, we flip it to `noop` by default.

**Comment by timdeschryver** [comment]
I like that suggestion @brandonroberts 👍

**Comment by arturovt** [comment]
Alright, so I should re-work the changes and add a new config property?

**Review by brandonroberts — APPROVED** [review]

File: modules/store-devtools/src/devtools.ts [hunk 1]
```diff
@@ -196,3 +218,26 @@ export class StoreDevtools implements Observer<any>, OnDestroy {
...
+  // with `next` being called within the Angular zone.
+  return (source: Observable<T>) =>
+    new Observable<T>((subscriber) =>
+      source.subscribe({
+        next: (value) => ngZone.run(() => subscriber.next(value)),
+        error: (error) => ngZone.run(() => subscriber.error(error)),
+        complete: () => ngZone.run(() => subscriber.complete()),
+      })
+    );
+}
```
Comment (markostanimirovic) [inline modules/store-devtools/src/devtools.ts:243]:
```suggestion
/**
 * If the devtools extension is connected out of the Angular zone,
 * this operator will emit all events within the zone.
 */
function emitInZone<T>(zoneConfig: ZoneConfig): MonoTypeOperatorFunction<T> {
  if (zoneConfig.connectInZone) {
    return (source) => source;
  }

  const { ngZone } = zoneConfig;

  return (source) =>
    new Observable<T>((subscriber) =>
      source.subscribe({
        next: (value) => ngZone.run(() => subscriber.next(value)),
        error: (error) => ngZone.run(() => subscriber.error(error)),
        complete: () => ngZone.run(() => subscriber.complete()),
      })
    );
}
```

File: modules/store-devtools/src/extension.ts [hunk 2]
```diff
@@ -80,7 +80,8 @@ export class DevtoolsExtension {
   constructor(
     @Inject(REDUX_DEVTOOLS_EXTENSION) devtoolsExtension: ReduxDevtoolsExtension,
     @Inject(STORE_DEVTOOLS_CONFIG) private config: StoreDevtoolsConfig,
-    private dispatcher: DevtoolsDispatcher
+    private dispatcher: DevtoolsDispatcher,
+    private ngZone: NgZone
```
Comment (markostanimirovic) [inline modules/store-devtools/src/extension.ts:84]:
```suggestion
    private dispatcher: DevtoolsDispatcher
```

Add private class member:

```ts
private zoneConfig: ZoneConfig;
```

in the constructor:

```ts
this.zoneConfig = injectZoneConfig(config.connectInZone);
```

File: modules/store-devtools/src/extension.ts [hunk 3]
```diff
@@ -167,10 +168,21 @@ export class DevtoolsExtension {
       return EMPTY;
     }
 
+    const ngZoneEnabled = this.config.ngZone === 'enabled';
+
```
Comment (markostanimirovic) [inline modules/store-devtools/src/extension.ts:172]:
```suggestion
```

File: modules/store-devtools/src/extension.ts [hunk 4]
```diff
@@ -167,10 +168,21 @@ export class DevtoolsExtension {
...
+        ? this.devtoolsExtension.connect(this.getExtensionConfig(this.config))
+        : // To reduce change detection cycles, we need to run the `connect` method
+          // outside of the Angular zone. The `connect` method adds a `message`
+          // event listener to communicate with an extension using `window.postMessage`
+          // and handle message events.
+          // We will re-enter the Angular zone when handling specific actions from the
+          // extension and when dispatching actions through `store.dispatch`.
+          this.ngZone.runOutsideAngular(() =>
+            this.devtoolsExtension.connect(this.getExtensionConfig(this.config))
+          );
```
Comment (markostanimirovic) [inline modules/store-devtools/src/extension.ts:184]:
```suggestion
      const connection = this.zoneConfig.connectInZone
        ? this.devtoolsExtension.connect(this.getExtensionConfig(this.config))
        : this.zoneConfig.ngZone.runOutsideAngular(() =>
            this.devtoolsExtension.connect(this.getExtensionConfig(this.config))
          );
```

File: modules/store-devtools/src/config.ts [hunk 5]
```diff
@@ -114,6 +114,13 @@ export class StoreDevtoolsConfig {
    * Maximum stack trace frames to be stored (in case trace option was provided as true).
    */
   traceLimit?: number;
+
+  /**
+   * The property determines whether the extension connection is established within the
+   * Angular zone or not. This is set to `enabled` by default, which means the connection
+   * is set up within the Angular zone by default to ensure backward compatibility.
+   */
+  ngZone?: 'enabled' | 'noop';
```
[the remark this answers is not in the document]
Comment (timdeschryver) [reply modules/store-devtools/src/config.ts:123]: I see why this cannot be very clear, and I agree that we should change it.
However, I don't really like to have a default value of `true` for a boolean.

What about the following?

```ts
(extension)ConnectionType?: 'ngZone' (default) | 'runOutsideAngular'
```

[the remark this answers is not in the document]
Comment (markostanimirovic) [reply modules/store-devtools/src/config.ts:123]: What about `connectOutsideZone` that would be `false` by default? 👀 

cc @brandonroberts @arturovt

[the remark this answers is not in the document]
Comment (timdeschryver) [reply modules/store-devtools/src/config.ts:123]: That's good for me 👍

[the remark this answers is not in the document]
Comment (brandonroberts) [reply modules/store-devtools/src/config.ts:123]: Seems reasonable to me

**Review by markostanimirovic — APPROVED** [review]
LGTM 👍 Thanks @arturovt!

**Comment by timdeschryver** [comment]
Thanks @arturovt !

**Comment by alex-okrushko** [comment]
Thanks for the fix @arturovt !
I ran into this issue myself, and when debugging I noticed it led back to ngrx/store-devtools 😱

@brandonroberts @markostanimirovic @timdeschryver I think it would make sense to have `runOutsideAngular` be enabled by default. I know it's a "change in behavior" but it's fixing something that should not have been there to begin with - app shouldn't rely on arbitrary change detection triggers outside of the app.

**Comment by brandonroberts** [comment]
> Thanks for the fix @arturovt ! I ran into this issue myself, and when debugging I noticed it led back to ngrx/store-devtools 😱
> 
> @brandonroberts @markostanimirovic @timdeschryver I think it would make sense to have `runOutsideAngular` be enabled by default. I know it's a "change in behavior" but it's fixing something that should not have been there to begin with - app shouldn't rely on arbitrary change detection triggers outside of the app.

Agreed. We can possibly land this change as part of the 17.x release and do a migration for existing apps also to keep the current behavior.

**Comment by markostanimirovic** [comment]
> Thanks for the fix @arturovt ! I ran into this issue myself, and when debugging I noticed it led back to ngrx/store-devtools 😱
> 
> @brandonroberts @markostanimirovic @timdeschryver I think it would make sense to have `runOutsideAngular` be enabled by default. I know it's a "change in behavior" but it's fixing something that should not have been there to begin with - app shouldn't rely on arbitrary change detection triggers outside of the app.

Sounds good to me 👍

**Pull request merged**